### DISCUSSION TITLE
use nested-object structured data with useFieldArray hook

### DISCUSSION THREAD

**Discussion by LeamonLee** [react-hook-form/react-hook-form#3117] [Q&A]
Hello,
I have a dynamic form which I need to render by reading a config file. The thing is since I would like to group related fields together so that I can organize them more easily, the structure of the form data is a nested object. But it can't seem to work well with useFieldArray hook. Is there any way that I can do something about it?

side note: I did see the console message saying the suggestion, but it's not the solution that I'm looking for.

The codesandbox link is provided below:
https://codesandbox.io/s/spring-silence-hloxb?file=/src/App.js

**Comment by bluebill1049** [comment]
I don't understand the problem that you try to resolve.

**Reply by LeamonLee** [reply]
The data structure of my form is a nested object instead of one layer hierarchy. Like this: 
{
   setting: {
      basic: {
        title: "test",
        keywords: ["testKeywords"],
        ...
      },
      common: {
        ...
      },
      advanced: {
       ...
      }
   }
}
However, useFieldArray doesn't work well with nested-object data structure when I use setValue function due to field name becomes 
settings.basic.keywords[0] 

<TextField
                inputRef={register()}
                type="input"
                name={`${fieldName}[${index}]`}
                id={`${fieldName}[${index}]`}
                label={label}
                variant="outlined"
                autoFocus
                fullWidth
                margin="normal"
                required
/>

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/tree/master/examples

search: Array of Field Array

**Reply by LeamonLee** [reply]
> https://github.com/react-hook-form/react-hook-form/tree/master/examples
> 
> search: Array of Field Array

Thanks for your prompt reply.
I just tried the way you suggested, but there's still a small issue about the initial value.
I set the initial value for the array field in useEffect function, and I printed in console by using getValues() method, but it seems that the initial value I tried to set doesn't correctly set to the array field. 
Like the following:

useEffect(() => {
    setValue("settings", {
      basic: {
        title: "test",
        // keywords: ["testKeywords"]
        keywords: {
          keywords: ["testKeywords"]
        }
      }
    });
  }, []);

But the console prints:
settings: {
    basic: {
        title: "test",
        keywords: {
           keywords: [""]
        }
    }
}
It's supposed to be like this:
settings: {
    basic: {
        title: "test",
        keywords: {
           keywords: ["testKeywords"]
        }
    }
}

The codesandbox link:
https://codesandbox.io/s/spring-silence-hloxb?file=/src/App.js:2607-2820
